### PR TITLE
Committee unit tests: constructor

### PR DESCRIPTION
**Pull request by pgev**
- Applies solidity style guide.
  - Applies js linter.
  - Introduces initial version for .travis.yml file.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  - sudo apt-get install nodejs
+  - sudo apt-get install npm
+  - sudo apt-get install software-properties-common
+  - sudo add-apt-repository -y ppa:ethereum/ethereum
+install:
+  - npm run update
+before_script:
+  - ./tools/run_ganache_cli.sh </dev/null 1>/dev/null 2>&1 &
+script:
+  - npm run test:contracts test/committee/constructor.js
```
Comment (benjaminbollen) [inline .travis.yml]: why is this explicit?

File: contracts/committee/Committee.sol [hunk 2]
```diff
@@ -17,77 +17,89 @@ pragma solidity ^0.5.0;
 import "../consensus/ConsensusModule.sol";
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 
-/**
- * @title Committee
- * @author Benjamin Bollen - <ben@ost.com>
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: I followed gnosis' style of listing authors

File: contracts/committee/Committee.sol [hunk 3]
```diff
@@ -17,77 +17,89 @@ pragma solidity ^0.5.0;
...
+    // @qn (pro): What is this for?
+    bytes32 public constant COMMIT_REJECT_PROPOSAL = keccak256(
+        "COMMIT_REJECT_PROPOSAL"
+    );
+
+
+    /* Enums */
+
+    /** Committee status enum. */
+    enum CommitteeStatus {
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol:80]: I feel it is not helpful to put so much whitespace in the definition of the enums

File: contracts/committee/Committee.sol [hunk 4]
```diff
@@ -98,57 +110,59 @@ contract Committee is ConsensusModule {
     uint256 public quorum;
 
     /**
-     * Committee members count
-     * counts members entered and reaches maximum at committeSize
+     * Committee members count.
+     *
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: the whitespace is redundant

File: contracts/committee/Committee.sol [hunk 5]
```diff
@@ -98,57 +110,59 @@ contract Committee is ConsensusModule {
...
+     * Counts members entered and reaches maximum at committeSize.
      */
-    uint256 public count; // TODO: rename memberCount
+    uint256 public memberCount;
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol:119]: thx for renaming

File: contracts/committee/Committee.sol [hunk 6]
```diff
@@ -98,57 +110,59 @@ contract Committee is ConsensusModule {
     uint256 public quorum;
 
     /**
-     * Committee members count
-     * counts members entered and reaches maximum at committeSize
+     * Committee members count.
+     *
+     * Counts members entered and reaches maximum at committeSize.
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: typo on my part `committeeSize`

File: contracts/committee/Committee.sol [hunk 7]
```diff
@@ -158,52 +172,78 @@ contract Committee is ConsensusModule {
...
-            "Only members can call this function.");
+
+    /** Modifiers */
+
+    modifier onlyMember() {
+        require(
+            members[msg.sender] != address(0),
+            "Only members can call this function."
+        );
+
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: no whitelines please

File: contracts/committee/Committee.sol [hunk 8]
```diff
@@ -158,52 +172,78 @@ contract Committee is ConsensusModule {
...
+    /* Special Functions */
 
     /**
-     * @notice Setup a new committee
+     * @notice Setups a new committee.
+     *
+     * @param _committeeSize Size of the committee. A minimum size is 3.
+     * @param _dislocation Used to dislocate a validator position in a hashed
+     *                     space before calculating a distance from a proposal
+     *                     for example. A non-zero value is expected.
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: without `for example`

File: contracts/committee/Committee.sol [hunk 9]
```diff
@@ -158,52 +172,78 @@ contract Committee is ConsensusModule {
...
+     * @param _proposal A proposal for committee to agree on. A non-zero value
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: `A proposal for the committee to evaluate. A non-zero value is required.`

File: contracts/committee/Committee.sol [hunk 10]
```diff
@@ -213,159 +253,225 @@ contract Committee is ConsensusModule {
...
         public
     {
-        require(_committeeSize >= 3,
-            "Committee size must not smaller than three.");
-        require(_dislocation != 0,
-            "Dislocation must not be zero.");
-        require(_proposal != 0,
-            "Proposal must not be zero.");
-        // ensure constants are for a strict majority (Remove)
-        assert(2 * COMMITTEE_SUPER_MAJORITY_NUMERATOR > COMMITTEE_SUPER_MAJORITY_DENOMINATOR);
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol:223]: please don't remove this without improving it

File: contracts/committee/Committee.sol [hunk 11]
```diff
@@ -213,159 +253,225 @@ contract Committee is ConsensusModule {
...
-        uint256 dBoundary = distance(shuffle(members[SENTINEL_MEMBERS]), proposal);
-        uint256 dExcludedMember = distance(shuffle(_excludedMember), proposal);
-        require(dExcludedMember < dBoundary,
-            "Member has been excluded only if distance to proposal is less than the furthest committee member.");
+        require(
+            members[_excludedMember] == address(0),
+            "Member should not already be in the committee."
+        );
+
+        uint256 dBoundary = distance(
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol]: here you should use `distanceToProposal(members[SENTINEL_MEMBER])` too then

**Review by benjaminbollen — COMMENTED** [review]
first comments

File: contracts/committee/Committee.sol (hunk 2 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Ok. Then we should update our solidity style guide and add there that developers should add their credentials in the following style. I will bring it back for now and will create a ticket in developer guidelines to add this.

File: contracts/committee/Committee.sol (hunk 3 above)
Comment (pgev) [reply contracts/committee/Committee.sol:80]: This follows our coding style guide: https://github.com/OpenST/developer-guidelines/blob/master/SOLIDITY_STYLE_GUIDE.md#blank-lines

File: contracts/committee/Committee.sol (hunk 4 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Ok.

File: contracts/committee/Committee.sol (hunk 10 above)
Comment (pgev) [reply contracts/committee/Committee.sol:223]: I've added unit test to cover this part. However, if there is better way to improve it please let me know. I will bring it back for now as per request.

Comment (benjaminbollen) [reply contracts/committee/Committee.sol:223]: a unit test is better, thanks

File: test/committee/utils.js [hunk 12]
```diff
@@ -39,20 +34,39 @@ function shuffleAccount(dislocation, account) {
...
+  Invalid: new BN(5),
+};
 
-  shuffleAccount,
+function isCommitteeOpen(status) {
+  return CommitteeStatus.Open.cmp(status) === 0;
+}
 
-  distance,
+const SENTINEL_MEMBERS = '0x1';
```
Comment (benjaminbollen) [inline test/committee/utils.js]: should this not be 20 bytes?

File: contracts/committee/Committee.sol (hunk 11 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Yes. I've grepped `distance(shuffle` ... which missed for this case as there is a new line. I will double check other places also.

**Review by benjaminbollen — COMMENTED** [review]
the travis.yml should be corrected

File: .travis.yml (hunk 1 above)
Comment (pgev) [reply .travis.yml]: Fixed.

File: contracts/committee/Committee.sol (hunk 2 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Here is the ticket in developer guidelines: https://github.com/OpenST/developer-guidelines/issues/33

I will revert back changes.

File: contracts/committee/Committee.sol (hunk 7 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Done.

File: contracts/committee/Committee.sol (hunk 8 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Done.

File: contracts/committee/Committee.sol (hunk 9 above)
Comment (pgev) [reply contracts/committee/Committee.sol]: Done.

File: test/committee/utils.js (hunk 12 above)
Comment (pgev) [reply test/committee/utils.js]: Yes. However, as it was not used (currently, we retrieve sentinel value from contract itself) I've removed it. Thx.

File: .travis.yml [hunk 13]
```diff
@@ -29,6 +29,7 @@ install:
 before_script:
   - ./tools/run_ganache_cli.sh </dev/null 1>/dev/null 2>&1 &
 script:
-  - npm run test:contracts test/committee/constructor.js
+  - npm run compile
+  - npm run test:contracts
```
Comment (benjaminbollen) [inline .travis.yml:33]: yay!

File: contracts/committee/Committee.sol [hunk 14]
```diff
@@ -18,7 +18,10 @@ import "../consensus/ConsensusModule.sol";
 import "openzeppelin-solidity/contracts/math/SafeMath.sol";
 
 
-/** @title Committee */
+/**
+ * @title Committee
+ * @author Benjamin Bollen - <ben@ost.com>
```
Comment (benjaminbollen) [inline contracts/committee/Committee.sol:23]: you should add Paruyr Gevorgyan now

File: test/committee/utils.js [hunk 15]
```diff
@@ -35,8 +43,8 @@ function shuffleAccount(dislocation, account) {
 
 function distance(h1, h2) {
   // Create BN from hashes.
-  const a = new BN(h1, 16);
-  const b = new BN(h2, 16);
+  const a = new BN(remove0x(h1), 16);
```
Comment (benjaminbollen) [inline test/committee/utils.js:46]: ah, nice!

File: package.json [hunk 16]
```diff
@@ -21,12 +27,14 @@
   "homepage": "https://github.com/openst/compute#readme",
   "devDependencies": {
     "assert": "^2.0.0",
-    "bn.js": "^4.11.8",
+    "bn.js": "^5.0.0",
+    "chai": "^4.2.0",
```
Comment (benjaminbollen) [inline package.json:31]: introducing chai and mocha introduces

```
npm i
added 85 packages from 314 contributors, updated 9 packages and audited 75844 packages in 4.151s
found 1597 vulnerabilities (2 low, 1595 high)
  run `npm audit fix` to fix them, or `npm audit` for details
```

I m not sure I want to use these

**Review by benjaminbollen — COMMENTED** [review]
Awesome. But I have questions around using chai or mocha

File: package.json (hunk 16 above)
Comment (benjaminbollen) [reply package.json:31]: forget my comment, it was already this bad

**Comment by benjaminbollen** [comment]
LGTM!

**Pull request merged**